packages/ddex/ingester/crawler/crawler.go

```diff
@@ -114,18 +120,283 @@ func (c *Crawler) pollS3Bucket(lastPolledTime time.Time) ([]*s3.Object, error) {
+	etag := strings.Trim(*upload.ETag, "\"")
+
+	// Only insert if a document doesn't already exist with this ETag
+	var delivery common.Delivery
+	filter := bson.M{
+		"_id": etag,
+	}
+	err := c.DeliveriesColl.FindOne(c.Ctx, filter).Decode(&delivery)
+	if err == nil {
+		c.Logger.Info("Skipping crawling delivery with ETag that already exists", "etag", etag)
+		return nil
+	}
+	if err != mongo.ErrNoDocuments {
+		c.Logger.Error("Error querying Mongo to check for existing ETag", "error", err)
+		return err
+	}
+
+	c.Logger.Info("Processing new ZIP file", "remotePath", remotePath, "etag", etag)
+
+	delivery = common.Delivery{
+		ZIPFileETag:    etag,
+		ZIPFilePath:    remotePath,
+		CreatedAt:      *upload.LastModified,
+		DeliveryStatus: constants.DeliveryStatusParsing,
+	}
+
+	// Download ZIP file from S3
+	zipFilePath, cleanup := c.downloadFromS3Raw(remotePath)
+	defer cleanup()
+	if zipFilePath == "" {
+		c.Logger.Error("Failed to download ZIP file from S3 bucket", "path", remotePath)
+		return err
+	}
+
+	// Unzip the file and crawl its contents
+	extractDir, err := os.MkdirTemp("", "extracted")
+	if err != nil {
+		c.Logger.Error("Error creating temp directory", "error", err)
+		return err
+	}
+	defer os.RemoveAll(extractDir)
+
+	if err := unzip(zipFilePath, extractDir); err != nil {
+		c.Logger.Error("Error unzipping file", "error", err)
+		return err
+	}
+
+	if c.DDEXChoreography == constants.ERNReleaseByRelease {
+		releases, err := c.crawlUnzippedReleaseByRelease(extractDir)
+		if err == nil {
+			delivery.Releases = *releases
+		} else {
+			delivery.ValidationErrors = append(delivery.ValidationErrors, "Error crawling unzipped release-by-release: "+err.Error())
+			delivery.DeliveryStatus = constants.DeliveryStatusErrorCrawling
+			c.Logger.Error("Error crawling unzipped release-by-release", "error", err)
+		}
+	} else {
+		batches, err := c.crawlUnzippedBatch(extractDir)
+		if err == nil {
+			delivery.Batches = *batches
+		} else {
+			delivery.ValidationErrors = append(delivery.ValidationErrors, "Error crawling unzipped batch: "+err.Error())
+			delivery.DeliveryStatus = constants.DeliveryStatusErrorCrawling
+			c.Logger.Error("Error crawling unzipped batch", "error", err)
+		}
+	}
+
+	// Insert the delivery into Mongo
+	_, err = c.DeliveriesColl.InsertOne(c.Ctx, delivery)
+	if err != nil {
+		c.Logger.Error("Error inserting delivery into Mongo", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// crawlUnzippedReleaseByRelease finds releases in unzippedRoot (and its subdirectories). For each release, assets are uploaded to the S3 "crawled" bucket, and metadata is stored in Mongo.
+// This follows "release-by-release" choreography: https://ernccloud.ddex.net/electronic-release-notification-message-suite-part-3%253A-choreographies-for-cloud-based-storage/5-release-by-release-profile/5.2-file-server-organisation/
+func (c *Crawler) crawlUnzippedReleaseByRelease(unzippedRoot string) (*[]common.UnprocessedRelease, error) {
```

Comment:
nit but would be nice for the release by release functions to be in their own file like you have for batch choreo with batch_choreography.go

Reply:
done